src/axl.c

```diff
@@ -57,14 +58,83 @@ Global Variables
  * before transferring files */
 int axl_make_directories;
 
-/* AXL's flush file, SCR has one as well */
-char* axl_flush_file = NULL;
+/*
+ * Array for all the AXL_Create'd kvtree pointers.  It's indexed by the AXL id.
+ *
+ * Note: We only expand this array, we never shrink it.  This is fine since
+ * the user is only going to be calling AXL_Create() a handful of times.  It
+ * also simplifies the code if we never shrink it, and the extra memory usage
+ * is negligible, if any at all.
+ */
+kvtree **axl_kvtrees;
+static unsigned int axl_kvtrees_count = 0;
+
+static pthread_mutex_t id_lock = PTHREAD_MUTEX_INITIALIZER;
+
+/*
+ * Allocate a new kvtree and return the AXL ID for it.  If state_file is
+ * specified, then populate the kvtree with it's data.
+ */
+static int axl_alloc_id(const char* state_file)
+{
+    kvtree *new;
+
+    new = kvtree_new();
+    if (state_file) {
+         /* initialize kvtree values from state_file if we have one */
+        kvtree_read_file(state_file, new);
+    } else {
+        /*
+         * No state file, just put in an empty string since you can't use
+         * a NULL for a string in kvtree.
+         */
+        state_file = "";
+    }
+
+    kvtree_util_set_str(new, AXL_KEY_STATE_FILE, state_file);
```

Comment:
It would be cleaner to avoid setting this field at all when the user passes state_file == NULL.  We can later check whether it was set with ``kvtree_util_get_str(kvtree, AXL_KEY_STATE_FILE, &state_file) == KV_TREE_SUCCESS)``.